### PR TITLE
🐛 Source Oura: Fix issues with date interpolation and datetime streams

### PR DESCRIPTION
**Pull request by isaacharrisholt**
## What

The Oura API does not allow requests made to the `heart_rate` endpoint to have a `start_datetime` and `end_datetime` more than 30 days apart. As this is the check stream, this is causing the checks to fail when those fields are more than 30 days apart, even if correctly formatted.

There is also an issue where the `start_date` and `end_date` parameters for the `date_requester` are being read from nonexistent config fields.

Relevant issues:

#19807 

## How

Implementing a stream slicer with 30 day increments for the `heart_rate` stream solved that problem, and the `date_requester` request options were corrected.

In order to add the stream slicer while being backwards compatible, I had to make an adjustment to the `day_delta` macro, adding an optional `from_date` parameter. This allowed me to set the default `start_datetime` to the `end_datetime` minus 1 day, regardless of what the `end_datetime` was set to.

I also added an `as_timezone` macro, which converts a `datetime` object or string to the specified timezone. This was needed to keep the `datetime_format` consistent while maintaining backward compatibility. Essentially, Python's `datetime.strptime` method won't read the `%z` format (timezone) if no timezone is present in the string:

```python
from datetime import datetime

datetime.strptime('2022-01-01T00:00:00', '%Y-%m-%dT%H:%M:%S%z')
# ValueError: time data '2022-01-01T00:00:00' does not match format '%Y-%m-%dT%H:%M:%S%z'
```

To combat this, I convert whatever datetime the user inputs to UTC, and force the stream slicer to use UTC format.

## Recommended reading order
1. `oura.yaml`
2. `spec.yaml`
3. `macros.py`
4. `test_macros.py`

## 🚨 User Impact 🚨

These changes will mean that streams using the `date_requester` will work as expected. Previously, they would only return data from the current day, but they will now return all data between the dates specified.

## Pre-merge Checklist

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

image

</details>

<details><summary><strong>Integration</strong></summary>

image

</details>

<details><summary><strong>Acceptance</strong></summary>

N/A

</details>

### CODE REVIEW

**Comment by marcosmarxm** [comment]
@brianjlai and @girarda can you check this contribution. It is doing a change in low-code CDK macros. Would be nice to get your feedback on that.

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/interpolation/macros.py [hunk 1]
```diff
@@ -83,17 +84,23 @@ def max(*args):
     return builtins.max(*args)
 
 
-def day_delta(num_days: int) -> str:
+def day_delta(num_days: int, from_date: Union[str, datetime.datetime] = None) -> str:
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/interpolation/macros.py:87]: nite: the type should be `Optional[Union[str, datetime.datetime]]`

**Review by girarda — APPROVED** [review]
me gusta! thank you for your contribution @isaacharrisholt

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/interpolation/macros.py (hunk 1 above)
Comment (isaacharrisholt) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/interpolation/macros.py]: Updated!

**Comment by marcosmarxm** [comment]
Hello :wave::skin-tone-2: and thank you for your contribution!

Airbyte has instituted a code freeze between 19 and 30 December, to make sure there are no disruptions during the holidays. 
Because of this, reviewing and merging your contribution may take longer than usual. 
We apologize for the delay, but we want everyone to have a quiet and happy holiday.

If you have any questions or need further clarification, please don't hesitate to ping via Slack.

**Comment by isaacharrisholt** [comment]
Hey, is there any update on when this is going to get merged?

**Comment by marcosmarxm** [comment]
@isaacharrisholt closing because there some conflicts and some updates in Airbyte CDK. Please submit the contribution again

**Pull request closed without merging**